### PR TITLE
chore(compass-schema-validation): convert to TS and new-style plugin COMPASS-7476

### PR DESCRIPTION
**Pull request by addaleax**
### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [x] Misc

## Types of changes

- [ ] *Backport Needed*
- [x] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-collection/src/components/collection-tab-provider.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+export interface CollectionTabPlugin {
+  name: string;
+  component: React.ComponentType<CollectionTabPluginMetadata>;
+}
+
+const CollectionTabsContext = React.createContext<CollectionTabPlugin[]>([]);
+
+export const CollectionTabsProvider: React.FunctionComponent<{
+  tabs: CollectionTabPlugin[];
+}> = ({ tabs, children }) => {
```
Comment (addaleax) [inline packages/compass-collection/src/components/collection-tab-provider.tsx:13]: @gribnoysup I saw you used `useRef()` for the tabs for the instance tabs provider, was there a specific reason for that? I wouldn’t assume that `tabs` change over the lifetime of the component, but if it _did_ change, it would probably do so for a reason, so wouldn’t we want to also change the value in the provider itself?

File: packages/compass-collection/src/components/collection-tab.tsx [hunk 2]
```diff
@@ -57,12 +62,35 @@ const collectionModalContainerStyles = css({
...
+  const legacyTabs = renderTabs();
+  const tabs = [
+    ...legacyTabs,
+    ...(localAppRegistry ? pluginTabs : []).map(
+      ({ name, component: Component }) => ({
+        name,
+        component: (
+          <AppRegistryProvider localAppRegistry={localAppRegistry}>
+            <Component {...collectionTabPluginMetadata} />
+          </AppRegistryProvider>
```
Comment (addaleax) [inline packages/compass-collection/src/components/collection-tab.tsx:89]: @gribnoysup ^ this probably looks wrong in some way? Wanted to get something minimal that seems to work fine to avoid conflicts with COMPASS-7402

File: packages/compass-schema-validation/src/components/sample-documents/sample-documents.spec.tsx [hunk 3]
```diff
@@ -1,23 +1,20 @@
...
 import { expect } from 'chai';
-import configureStore from '../../stores';
-import SampleDocuments from '../sample-documents';
+import SampleDocuments from '.';
 import { Provider } from 'react-redux';
+import { createBareStore } from '../../stores/store';
 
 describe('SampleDocuments [Component]', function () {
   it('renders a valid and invalid document preview', function () {
+    const store = createBareStore();
```
Comment (addaleax) [inline packages/compass-schema-validation/src/components/sample-documents/sample-documents.spec.tsx:10]: Might not be a pattern we’d really want to adopt, but this was just very easy for converting the existing code

File: packages/compass-schema-validation/src/modules/index.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  | DataServiceAction
+  | FieldsAction
+  | NamespaceAction
+  | ServerVersionAction
+  | ValidationAction
+  | SampleDocumentAction
+  | IsZeroStateAction
+  | IsLoadedAction
+  | EditModeAction
+  | ResetAction;
```
Comment (addaleax) [inline packages/compass-schema-validation/src/modules/index.ts:67]: (also aware that it probably makes sense to put much of these into a single reducer, just wanted to do the most direct/mechanical translation here)

File: packages/compass-schema-validation/src/modules/sample-documents.ts [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    const state = getState();
+    const dataService = state.dataService;
+    const namespace = state.namespace.ns;
+    const validator = state.validation.validator;
+
+    // Calling checkValidator twice here seems wrong
+    const checkedValidator = checkValidator(validator);
+    const query = checkValidator(
+      checkedValidator.validator as string
+    ).validator;
```
Comment (addaleax) [inline packages/compass-schema-validation/src/modules/sample-documents.ts:287]: Added the comment here, something can’t be quite right here but the code was admittedly already a bit hard to follow because it somewhat freely mixes validators in string representation vs JS object representation

File: packages/compass-schema-validation/src/stores/store.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+  if (options.namespace) {
+    const namespace = toNS(options.namespace);
+    store.dispatch(namespaceChanged(namespace));
+  }
+
+  // XXX: Verify that this is the correct way of handling this now that tabs are loaded ad-hoc?
+  (store.dispatch as ThunkDispatch<RootState, unknown, RootAction>)(
+    activateValidation()
+  );
```
Comment (addaleax) [inline packages/compass-schema-validation/src/stores/store.ts:109]: @gribnoysup See the comment, this seems reasonable as a dispatched action, do we just call it in the `activate()` method?

File: packages/compass-collection/src/components/collection-tab-provider.tsx (hunk 1 above)
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection-tab-provider.tsx:13]: I'm assuming we always want the provider to work as initial configuration for tabs and using a ref removes the need to use memo when rendering a provider. We can change that if you feel like it and to be consistent probably better to do it on other places too them, just need to adjust other places where we already have providers like that to account for that

File: packages/compass-collection/src/components/collection-tab.tsx (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection-tab.tsx:89]: Yeah, the scope for the tab should be controlled by the parent, not the tab itself, this is why this localAppRegistry state is already part of the store in the tabs store, not in the collection-tab store

File: packages/compass-schema-validation/src/stores/store.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import type { DataService } from 'mongodb-data-service';
+import type { MongoDBInstance } from '@mongodb-js/compass-app-stores/provider';
+
+/**
+ * The lowest supported version.
+ */
+const MIN_VERSION = '3.2.0';
+
+// Exposed for testing
+export function createBareStore(): Store<RootState, RootAction> {
```
Comment (gribnoysup) [inline packages/compass-schema-validation/src/stores/store.ts:30]: Exporting it for testing is fine, but we do call those methods just `configureStore` usually

File: packages/compass-schema-validation/src/modules/index.ts (hunk 4 above)
Comment (gribnoysup) [reply packages/compass-schema-validation/src/modules/index.ts:67]: That's fine, yeah, I just personally find it less work to rewrite to one file instead of converting file-by-file

File: packages/compass-schema-validation/src/stores/store.ts (hunk 6 above)
Comment (gribnoysup) [reply packages/compass-schema-validation/src/stores/store.ts:109]: I don't see anything wrong with that. If it's "sync" state, you can pass it to the configure store, if it's async action that needs to run when the plugin is first activated, you can dispatch it in the activate method

File: packages/compass-collection/src/components/collection-tab.tsx (hunk 2 above)
Comment (addaleax) [reply packages/compass-collection/src/components/collection-tab.tsx]: @gribnoysup How does 92c0d6afbdcee14fc104dbe33c024469fa3c4715 look to you?

File: packages/compass-collection/src/components/collection-tab-provider.tsx (hunk 1 above)
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection-tab-provider.tsx:13]: > we’re not changing tabs lists at runtime

That's kinda my point, if we don't ref this value we're now need to keep in mid that passing values there in runtime will actually change it and will rerender children using this context unneccessarily if we're not careful, making you think more about it, not less, which I'm trying to avoid. As I said though, feel free to update this, just also need to move other tab props outside of the render / memo them so that we don't actually update the value in runtime

File: packages/compass-schema-validation/src/stores/store.ts (hunk 7 above)
Comment (gribnoysup) [reply packages/compass-schema-validation/src/stores/store.ts]: Ahhh, gotcha, sorry, I think I just didn't notice that the `configureStore` was still there

File: packages/compass-collection/src/modules/tabs.ts [hunk 8]
```diff
@@ -213,8 +213,11 @@ const createNewTab = (
       localAppRegistry,
       ...collectionMetadata,
     });
-    const component = React.createElement(collectionTabRole.component, {
-      store,
+    const component = React.createElement(AppRegistryProvider, {
+      localAppRegistry: localAppRegistry,
```
Comment (gribnoysup) [inline packages/compass-collection/src/modules/tabs.ts:217]: Yep, that looks good! One thing here still needs handling, as we discussed in the tech design, this is finally a case for us to use `deactivateOnUnmount` prop and disable it, otherwise switching to some other workspace, like instance for example, cleans up `localAppRegistry` for the tab

File: packages/compass-collection/src/components/collection-tab.tsx (hunk 2 above)
Comment (addaleax) [reply packages/compass-collection/src/components/collection-tab.tsx]: Actually … with that change, something odd that’s happening is that the top-level tab titles don’t update if I switch to another tab (Indexes is the one in the failing e2e test here) if and only if I open a collection through the collections list view, doesn’t happen when I do it through the sidebar … I’ll probably figure it out, but if this rings a bell to you, lmk :)

Comment (addaleax) [reply packages/compass-collection/src/components/collection-tab.tsx]: Ah nvm, `deactivateOnUnmount` did the trick I think :) But tbh I’m a bit surprised that _that_ is where the visible difference between having the option and not having it falls …

**Review by gribnoysup — APPROVED** [review]

File: packages/compass-collection/src/components/collection-tab.tsx (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-collection/src/components/collection-tab.tsx]: This state does rely on localAppRegistry events at the moment 😞, as you as you switch the tab, this subscription was lost forever (a less immediately obvious thing was that scoped modals like explain plan would stop working after the first tab was opened and switched)
